### PR TITLE
add nvme reservation support

### PR DESCRIPTION
**Pull request by yocalebo**

### CODE REVIEW

File: nvme.pyx [hunk 1]
```diff
@@ -53,3 +121,156 @@ def get_nsid(path):
...
+
+cdef class NvmeDevice(object):
+
+    cdef const char *dev
+    cdef int fd
+
+    def __cinit__(self, path):
+
+        self.dev = path
+        self.fd = open(self.dev, O_RDONLY)
```
Comment (william-gr) [inline nvme.pyx]: Wonder if open/close could block for some? If so we would need to release GIL

**Comment by william-gr** [comment]
@amotin since you are familiar with the C code, could you please help us review this?

File: nvme.pyx (hunk 1 above)
Comment (yocalebo) [reply nvme.pyx]: The entire purpose for me brining in fcntl standard open and close functions was to release the gil when opening the device. I just simply forgot to add those parameters. Great catch, I'll update.

**Comment by yocalebo** [comment]
Tested this all weekend on an m-series. Setting/updating/changing reservation keys for nvme drives is working without issues and is stable.

**Comment by william-gr** [comment]
@amotin @sonicaj ping

**Comment by amotin** [comment]
Looking and working with @yocalebo .

**Review by amotin — APPROVED** [review]
Makes sense to me now.

File: nvme.pyx [hunk 2]
```diff
@@ -25,19 +25,79 @@
...
-from libc.stdint cimport uint32_t
 import os
 from posix.ioctl cimport ioctl
+from posix.strings cimport bzero
+from libc.string cimport memset
+from libc.stdint cimport uint8_t, uint16_t, uint32_t, uint64_t
 
 
-cdef extern from "dev/nvme/nvme.h":
+cdef int min(int a, int b):
```
Comment (sonicaj) [inline nvme.pyx]: I believe we can use the builtin `min` instead of defining this one ?

File: nvme.pyx [hunk 3]
```diff
@@ -53,3 +113,206 @@ def get_nsid(path):
...
+
+        n = min(n, (sizeof(data) - sizeof(s)) // sizeof(s.ctrlr[0]))
+
+        controller = {}
+        for i in range(n):
+            controller['controller_id'] = s.ctrlr[i].ctrlr_id
+            controller['resv_status'] = s.ctrlr[i].rcsts
+            controller['host_id'] = s.ctrlr[i].hostid
+            controller['key'] = s.ctrlr[i].rkey
+            info['controllers'].append(controller.copy())
```
Comment (sonicaj) [inline nvme.pyx]: How about
```
info['controllers'].append({"key": "value", "key": "value"})
```
`.copy` will work in this case but `.copy` does not make a deep copy and the errors can be surprising then ; )

File: nvme.pyx [hunk 4]
```diff
@@ -53,3 +113,206 @@ def get_nsid(path):
...
+            2. the type of reservation that
+                is being held (scopetype)
+            3. the reservation key that reserves
+                the disk (reservation)
+        """
+
+        data = self.resvreport()
+
+        for i in data['controllers']:
+            if i.get('resv_status') == 1:
```
Comment (sonicaj) [inline nvme.pyx]: Is `resv_status` ever going to not exist ? I think no, in this case we should access it directly so it fails we know something is off.

**Review by sonicaj — COMMENTED** [review]
LGTM, just some minor nits

File: nvme.pyx [hunk 5]
```diff
@@ -53,3 +108,208 @@ def get_nsid(path):
...
+        self.dev = path
+        with nogil:
+            self.fd = open(self.dev, O_RDONLY)
+            if self.fd == -1:
+                raise OSError(f'Failed to open: {self.dev}')
+
+    def __dealloc__(self):
+
+        if self.fd != -1:
+            close(self.fd)
```
Comment (sonicaj) [inline nvme.pyx]: `with nogil` ?

Comment (yocalebo) [reply nvme.pyx]: indeed

**Review by sonicaj — APPROVED** [review]

**Pull request merged**